### ISSUE TITLE
Adding type to container and exposing the type with IServiceProvider, after configuring the root container

### ISSUE THREAD

**Issue by jan-johansson-mr** [z4kn4fein/stashbox-extensions-dependencyinjection#11]
Hello,

I'm testing stashbox and it looks good. I want to add types (by later loading of modules) to the type container and make (the later) added types available with the service provider - and it seems to be working fine with stashbox. I already tested with Lamar with success, but the author thinks that you should not be able to add types to the underlying service provider, once you have built the provider. However, this is possible via some old code path that still remains in Lamar, accessible via its API.

Here is (some of) the code that I've been using while testing stashbox

```
    private static void StashboxTest()
    {
        var container = new StashboxContainer();

        container.RegisterInstance<IStashboxContainer>(container);
        container.Register<IX, A>(options => options.WithScopedLifetime());
        container.Validate();

        var p0 = new StashboxServiceProviderFactory(container);
        var q0 = p0.CreateServiceProvider(container);

        TestWithServiceProvider(q0);
    }

    static void TestWithServiceProvider(IServiceProvider serviceProvider)
    {
        var container = serviceProvider.GetRequiredService<IStashboxContainer>();

        container.Register<IY, B>(options => options.WithScopedLifetime());

        var p0 = serviceProvider.GetRequiredService<IX>();
        var p1 = serviceProvider.GetRequiredService<IY>();

        p1.DoWork();
    }
```

So what is happening with the code is that I create the container, then register type `A` with interface `IX`, and by the extension builds the service provider. Then I access the original container in the test call and there I register type `B` with interface `IY`. It happens that type `A` is dependent on `IY` via constructor. The code works, and I'm happy.

However, is this a feature that you intend for stashbox? I ask this out of consideration of what the author said about Lamar.

Thanks

**Comment by z4kn4fein** [comment]
Hi @jan-johansson-mr, thank you for reaching out!

This behavior is intended, yes. The container behind the service provider does not have a specific state where you are not allowed to do registrations. This has some tradeoffs though:
- Service resolutions build a delegate cache to speed up object creation for future requests. Each time a registration happens, the container wipes out that delegate cache, so services from that point on can be constructed with dependencies from the most up-to-date service registry. Without a built-up cache, resolution requests can be slower and more resource hungry. So, it's recommended not to register services "on the fly" too frequently. If this is not an issue for you or does not happen often, then it should be fine. But it is worth keeping this behavior in mind. :)
- Already constructed singletons and scoped services are not wiped out (as they must be created once), so they will remain in the current scope (and singleton scope) built with their original dependencies.

In your example, these do not seem to happen anyway, as you don't resolve anything before the "post service provider" registrations.

Feel free to let me know if you have any more questions!

**Comment by jan-johansson-mr** [comment]
Hello @z4kn4fein, thanks for your answer.

I did dive into the code, and indeed understood that it's possible to add to the container as-you-go. This is a good thing, since lazy loading of dependencies is needed in certain scenarios.

Thanks for pointing out the cons and pros of dependencies lazy loading. I'll be sharing my scenario for context, and because it's fun. My scenario is lazy loading with Blazor, using the pattern micro frontend. Using the lazy loading pattern with the "standard" dependency injection framework is simply not possible, since the framework assumes a static registration of all possible services at startup. So I found Lamar DI and then could do what I wanted. However, Lamar felt like it can break the lazy loading of dependencies at any updated version, so I found your DI, and I feel good about the solution you have.
